File: bin/dbt-plot-sar-swap.jl
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2015      Mark Wong
#               2015      2ndQuadrant, Ltd.
#

#=
INPUTFILE=$1
OUTPUTDIR=$2

if [ $# -ne 2 ]; then
	echo "Create a plot of swap statistics."
	echo "usage: $(basename "$0") <sar-swap.csv> <output directory>"
	echo
	echo "    <sar-swap.csv> - full path to the sar-net.csv file"
	echo "    <output directory> - location to write output files"
	echo
	echo "Will attempt to create <output directory> if it does not exist."
	exit 1
fi

exec julia --color=no --startup-file=no "$0" "$@"
=#

using CSV
using DataFrames
using VegaLite

function load(params)
    df = CSV.File(params["filename"]) |> DataFrame

    # Convert timestamp to elapsed time in minutes, in-place.

    start_time = minimum(df.timestamp)
    df = df[!, Not(["# hostname", "interval"])]
    transform!(
            df,
            :timestamp => ByRow(x -> (x - start_time) / 60) => :timestamp;
    )

    return df
end

function plot(df, params)
    idf = df[!, ["timestamp", "pswpin/s"]]
    rename!(idf, ["timestamp", "pswp/s"])
    insertcols!(idf, :label => "pswpin/s")

    odf = df[!, ["timestamp", "pswpout/s"]]
    rename!(odf, ["timestamp", "pswp/s"])
    insertcols!(odf, :label => "pswpout/s")
    idf = vcat(idf, odf)

    p = idf |>
        @vlplot(
            title="Swap Activity",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :timestamp,
                title="Elapsed Time (minutes)",
            },
            y={
                field="pswp/s",
                title="Pages/s",
            },
            color={
                :label,
                title="",
            },
        )
    filename = params["outputdir"] * "/sar-swap.png"
    save(filename, p)
end

function main()
    params = Dict(
            "filename" => ARGS[1],
            "outputdir" => ARGS[2],
    )

    df = load(params)
    plot(df, params)
end

main()

File: bin/dbt-plot-sar-swap.r
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright The DBT Tools Authors
#

INPUTFILE=$1
OUTPUTDIR=$2

if [ $# -ne 2 ]; then
	echo "Create a plot of swap statistics."
	echo "usage: `basename $0` <sar-swap.csv> <output directory>"
	echo
	echo "    <sar-swap.csv> - full path to the sar-net.csv file"
	echo "    <output directory> - location to write output files"
	echo
	echo "Will attempt to create <output directory> if it does not exist."
	exit 1
fi

R --slave --no-save << __EOF__
df <- read.csv("$INPUTFILE", sep=";", header=T)
starttime = df[1,]\$timestamp
df\$timestamp <- ceiling((df\$timestamp - starttime) / 60)

# Plot pages swapped in/out.

label <- c('pswpin/s', 'pswpout/s')
color <- rainbow(2)
pch <- c(1, 2)
bitmap(paste("$OUTPUTDIR/sar-swap.png", sep=""),
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$timestamp, df\$pswpin.s,
     ylim=c(0, max(max(df\$pswpin.s), max(df\$pswpout.s))),
     type = "b", col = color[1],
     main=paste("Swap Activity", sep=""),
     xlab="Elapsed Time (minutes)", ylab="Pages/s")
points(df\$timestamp, df\$pswpout.s, type = "b", pch = pch[2],
       col=color[2])

legend('topright', label, pch=pch, col=color)

grid(col="gray")
invisible(dev.off())
__EOF__
